armi/utils/monkeyPatcher.py

```diff
@@ -0,0 +1,162 @@
+active and will pull in the custom code as specified in the patch file:
+  - applyPreOpPatch
+  - applyPostOpPatch
+  - applyPostInterfacePatch
+  - applyPostRestartLoadPatch
+
+The easiest way to understand the hook locations is to grep the codebase for
+the calls listed above.
+
+If the patchFilePath parameter is not specified in the ARMI input yaml, the
+hooks will return immediately.
+
+The format of a patch file is almost entirely left to the end user, however
+there are some examples in the test module for anticipated uses of this module.
+Specifically, how to modify objects in higher levels of scope.
+
+Input
+-----
+The module requires a valid input file located at "patchFilePath" that 
+contains the functions:
+  - preOpPatch
+  - postOpPatch
+  - postInterfacePatch
+  - postRestartLoadPatch
+
+Each of these functions MUST be defined in the patchfile, even if they simply
+return on call.
+
+These functions are called in armi.cases.case.py and 
+armi.bookkeeping.mainInterface.py. In the event that no patch file path was
+specified, the functions in this module will return immediately. The logic
+was handled here in order to keep changes to the main armi workflow low.
+"""
+import os
+import importlib.util
+
+from armi.settings import caseSettings
+from armi.utils import getFileSHA1Hash
+
+# TODO: add logging
+
+
+def checkPatchFlag(f):
+    """
+    This checks if the patchFlag is True. This is used to build in a short-
+    circuit for the patcher, where the patcher will return to the main ARMI
+    process if no patchfile was provided.
+
+    The reason for including this is to keep modifications to the main ARMI
+    code to a minimum, so this effectively handles the check of whether there
+    is a patch to apply or not. This is formatted as a decorator since the
+    operation is applied to all the patch hooks.
+    """
+
+    def newfunc(*args, **kwargs):
+        self_arg = args[0]
+        if self_arg.patchFlag:
+            return f(*args, **kwargs)
+        else:
+            return
+
+    return newfunc
+
+
+class Patcher:
+    """
+    A controller object for patching ARMI at runtime. See module docstring for
+    instructions on usage and patch file formatting.
+
+    Input
+    -----
+    cs : armi.settings.caseSettings
+        The case settings for the ARMI job. The path to the patch file is stored
+        in the case settings.
+    """
+
+    def __init__(self, cs: caseSettings):
+        self.patchPath = cs["patchFilePath"]
+        self.patchFlag = True
+        if self.patchPath == "":
+            # TODO: Log that no patch file was provided.
+            self.patchFlag = False
+            return
+        if not os.path.isfile(self.patchPath):
+            self.patchFlag = False
+            raise IOError("The provided path to the patch file is invalid.")
+        self.hash = getFileSHA1Hash(self.patchPath)
+        # TODO: log the hash somewhere.
```

Comment:
Our goal, generally, is to not add new `TODO`s to the code.

https://terrapower.github.io/armi/developer/standards_and_practices.html#general-do-s-and-don-ts

Certainly, these all seem like easy `TODO`s to solve.